torch/_tensor_str.py

```diff
@@ -138,178 +139,72 @@ def _number_format(tensor, min_sz=-1):
-                strt += __repr_row(row, indent, fmt, scale, sz, n)
-        elif has_vdots and not has_hdots:
-            vdotfmt = u"{:^" + \
-                str(len(__repr_row(self[0], '', fmt, scale, sz))) + \
-                "}\n"
-            for row in self[:n]:
-                strt += __repr_row(row, indent, fmt, scale, sz)
-            strt += vdotfmt.format(u'\u22EE')
-            for row in self[-n:]:
-                strt += __repr_row(row, indent, fmt, scale, sz)
-        else:
-            for row in self:
-                strt += __repr_row(row, indent, fmt, scale, sz)
-    return strt
-
-
-def _vector_str(self):
-    fmt, scale, sz = _number_format(self)
-    strt = ''
-    ident = ''
-    n = PRINT_OPTS.edgeitems
-    dotfmt = u"{:^" + str(sz) + "}\n"
-    if scale != 1:
-        strt += SCALE_FORMAT.format(scale)
-        ident = ' '
-    if self.numel() < PRINT_OPTS.threshold:
-        return (strt +
-                '\n'.join(ident + fmt.format(val.item() / scale) for val in self) +
-                '\n')
+        data = [fmt.format(val.item() / scale) for val in self]
+
+    data_lines = [data[i:i + elements_per_line] for i in range(0, len(data), elements_per_line)]
+    lines = [','.join(line) for line in data_lines]
+    return '[' + (',' + '\n' + ' ' * (indent + 1)).join(lines) + ']'
+
+
+def _tensor_str(self, indent, fmt, scale, sz):
+    dim = self.dim()
+
+    if dim == 0:
+        return _scalar_str(self)
+    if dim == 1:
+        return _vector_str(self, indent, fmt, scale, sz)
+
+    if self.size(0) > PRINT_OPTS.threshold:
+        slices = ([_tensor_str(self[i], indent + 1, fmt, scale, sz) for i in range(0, PRINT_OPTS.edgeitems)] +
+                  ['...'] +
+                  [_tensor_str(self[i], indent + 1, fmt, scale, sz) for i in range(len(self) - PRINT_OPTS.edgeitems,
+                                                                                   len(self))])
     else:
-        return (strt +
-                '\n'.join(ident + fmt.format(val.item() / scale) for val in self[:n]) +
-                '\n' + (ident + dotfmt.format(u"\u22EE")) +
-                '\n'.join(ident + fmt.format(val.item() / scale) for val in self[-n:]) +
-                '\n')
+        slices = [_tensor_str(self[i], indent + 1, fmt, scale, sz) for i in range(0, self.size(0))]
+
+    tensor_str = (',' + '\n' * (dim - 1) + ' ' * (indent + 1)).join(slices)
+    return '[' + tensor_str + ']'
 
 
-def _str(self, include_footer=True):
+def _str(self):
     if self.is_sparse:
         size_str = str(tuple(self.shape)).replace(' ', '')
         return '{} of size {} with indices:\n{}and values:\n{}'.format(
             self.type(), size_str, self._indices(), self._values())
 
-    empty = self.numel() == 0
-    dim = self.dim()
+    type_str = 'tensor'
+    prefix = type_str + '('
+    indent = len(prefix)
+    if str(self.dtype) not in IMPLICIT_DTYPES:
+        suffix = ', dtype=' + str(self.dtype) + ')'
```

Comment:
A drawback of adding the shape representation in that case is that our `repr` doesn't match the constructor anymore.
But the same happens in numpy:
```python
print(np.zeros((0, 4)))
# returns
# array([], shape=(4, 0), dtype=float64)

np.array([], shape=(4, 0), dtype=np.float64)
# raises
# 'shape' is an invalid keyword argument for this function
```

Reply:
Okay sure.

I think it's fine that it doesn't match the constructor, it already doesn't in a few cases.

Reply:
Actually, is it possible to create a tensor with self.numel() == 0 that isn't just shape (0,)? Our constructors just parse the args sequentially and stop when they stop making sense. If we can't have non-zero dims in an empty tensor, I don't see a reason to add the shape into the repr.

```
>>> torch.zeros(2,0)
tensor([ 0, 0])
>>> torch.zeros(0,2)
tensor([], shape=(0,))
```

Reply:
Last time I checked, we could have something like `torch.Tensor(2, 0)` be an empty tensor, not sure it changed in the meantime

Reply:
Yeah it just makes a size 2 tensor now.

Reply:
I think I have mixed something up in my mind.
Still, I believe we will be extending support for tensors with 0 dimensions at some point in time (is that right @gchanan ?), so it might still be useful to have the case of `tensor.numel() == 0 and tensor.ndimension() > 0` to print the `shape` as well?